shared/src/pubsub/publisher.rs

```diff
@@ -18,12 +30,43 @@ impl Publisher {
             .wait_for_connect()
             .await
             .map_err(PublisherError::InitialConnection)?;
-        Ok(Self { client })
+        Ok(Self {
+            client,
+            rate_limiter: Arc::new(RateLimiter::keyed(Quota::per_minute(
+                NonZeroU32::new(RATE_LIMIT).unwrap(),
+            ))),
+        })
+    }
+
+    pub fn rate_limit_publisher(
+        &self,
+        key: &'static str,
+    ) -> Result<&Publisher, NotUntil<QuantaInstant>> {
+        self.rate_limiter.check_key(&key)?;
+        Ok(self)
+    }
```

Comment:
```suggestion

```

Reply:
Your earlier suggestion was right. This should not be public.